lib/Backend/Vault/Kv.hs

```diff
@@ -0,0 +1,234 @@
+                    )
+
+          fields <- maybeThrow $
+            secrets & traverse %~ (keyToField . fst) & sequence <&> HS.fromList
+          _tags <- maybeThrow $ cofferSpecials ^. tags & mapM E.newEntryTag
+
+          pure $ E.newEntry path (cofferSpecials ^. globalDateModified)
+            & E.masterField .~ (cofferSpecials ^. masterKey >>= E.newFieldKey)
+            & E.fields .~ fields
+            & E.tags .~ _tags
+      ListSecrets path ->
+        runMaybe (embedCatchMaybe path (listSecrets env path) <&> (^. I.ddata) <&> \(I.ListSecrets list) -> list)
+      DeleteSecret path ->
+        embedCatch path (deleteSecret env path) <&> const ()
+  where postSecret env = (I.routes env ^. I.postSecret) mount token
+        readSecret env = (I.routes env ^. I.readSecret) mount token
+        listSecrets env = (I.routes env ^. I.listSecrets) mount token
+        updateMetadata env = (I.routes env ^. I.updateMetadata) mount token
+        deleteSecret env = (I.routes env ^. I.deleteSecret) mount token
+
+        exceptionHandler :: [T.Text] -> ClientError -> Maybe CofferError
+        exceptionHandler path =
+          \case FailureResponse _request response ->
+                    case statusCode $ responseStatusCode response of
+                      404 -> Nothing
+                      e -> Just $ OtherError (T.pack $ show e)
+                DecodeFailure text response -> Just MarshallingFailed
+                UnsupportedContentType mediaType response -> Just MarshallingFailed
+                InvalidContentTypeHeader response -> Just MarshallingFailed
+                ConnectionError excepton -> Just ConnectError
+        embedCatch :: Member (Embed IO) r
+                       => Member (Error CofferError) r
+                       => [T.Text]
+                       -> IO a
+                       -> Sem r a
+        embedCatch path io = embed (catch @ClientError (io <&> Left) (pure . Right . exceptionHandler path)) >>=
+          \case Left l -> pure l
+                Right (Just r) -> throw r
+                Right Nothing -> throw $ OtherError "404"
+
+        embedCatchMaybe :: Member (Embed IO) r
+                        => Member (Error CofferError) r
+                        => Member (Error ()) r
+                        => [T.Text]
+                        -> IO a
+                        -> Sem r a
+        embedCatchMaybe path io = embed (catch @ClientError (io <&> Left) (pure . Right . exceptionHandler path)) >>=
+          \case Left l -> pure l
+                Right (Just r) -> throw r
+                Right Nothing -> throw ()
+
+        runMaybe :: Sem (Error () ': r) t
+                 -> Sem r (Maybe t)
+        runMaybe x = runError x <&> \case
+                       Left _ -> Nothing
+                       Right a -> Just a
+
+        maybeThrow :: Member (Error CofferError) r
+                      => Maybe a
+                      -> Sem r a
+        maybeThrow = maybe (throw MarshallingFailed) pure
```

Comment:
There 4 should be in some `Util.hs`.

Why a function that is called `maybeThrow` throws domain-related exception? Should be abstracted out.

Reply:
The name may be bad, it's supposed to convey: "If Nothing, throw this specific error specific to this module" It's just a helper to make the code a tad more readable.

Reply:
In that case, I suggest making it 
```hs
orThrow :: Member (Error e) r => Maybe a -> e -> Sem r a
```
after merge, which will allow using it in various other places.

This order of arguments would make it easy to use it as operator.
```hs
mbSomething `orThrow` MarshallingFailed
```

Reply:
Done